### ISSUE TITLE
allRules = true has no effect when buildUponDefaultConfig = true

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#4926]
## Expected Behavior
One of:
 * `allRules` enables all rules on top of using the default config (which seems to be the code (`AllRulesConfig`) anyway.
 * A build failure is triggered because the CLI throws an exception that conflicting flags were passed in.

## Observed Behavior
When configuring `DetektExtension`
```
buildUponDefaultConfig = true
allRules = true
```
bahaves the same as
```
buildUponDefaultConfig = true
```

## Steps to Reproduce
Will produce if necessary.

## Context
I want to go all in, making a reusable strictest configuration I can make, by enabling all checks, but keeping it open that any new Detekt improvements are automatically added.

## Your Environment

* Version of detekt used: 1.20.0
* Version of Gradle used (if applicable): 7.4.2
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: Windows 10
* Link to your project (if it's a public repository): N/A

**Comment by BraisGabin** [comment]
I was looking at this issue. And I found tests that asserts that exactly the behaviour that you describes. If I "fix" this issue I would break those tests. I agree that the behaviour is strange but I don't want to expend more time on this.

If anyone want to go deep in the rabbit hole and find out why we have that behaviour I would happy to read the findings. The configuration of detekt is something that should be rewritten for 2.0.

My findings:
- The test `should override active when specified` from `WorkaroundConfigurationKtSpec` defines the curren behaviour.
- An empty config `yml` and provide no config at all have different behaviours 🤷

My two cents: We should probably remove `allRules`. #5089 helps out to keep your configuration up to date with the last updates of detekt and you can always enable all the rules there. And it gives you more flexibility to disable rules that you don't like.

**Comment by TWiStErRob** [comment]
Ugh, that's not good, I do not want to list all the rules detect has, it's just noise in the configuration. It's much simpler to opt in to `allRules`, and disable whatever is not necessary. It's a lot of maintenance burden to add every new rule on every update rather than getting them "for free" on the cost of updating a number.

Compare [default](https://github.com/detekt/detekt/blob/main/detekt-core/src/main/resources/default-detekt-config.yml) (as an example of a config that list "almost" everything) vs [my opt out](https://github.com/TWiStErRob/net.twisterrob.sun/blob/master/config/detekt/detekt.yml) configs.

Note: Android lint has the same config and it survived many majors, because it's a good config option.   See also an example [lint config](https://github.com/TWiStErRob/net.twisterrob.sun/blob/master/config/lint/lint.xml) which opts out of "enable all warnings".

**Comment by TWiStErRob** [comment]
This might change if https://github.com/detekt/detekt/issues/4925 is possible... because then creating the new config is not tedious.

**Comment by mustafaozhan** [comment]
Interested +1

**Comment by TWiStErRob** [comment]
Which of the two expected behaviours would you expect, @mustafaozhan (and any other watchers, if any)?

**Comment by mustafaozhan** [comment]
My understanding should be as below:

- `buildUponDefaultConfig` -> Enables all the default `detekt` configurations.
- `allRules` -> Enables the experimental, upcoming new rules.

So either of the flags shouldn't override the other one and if a user want's to all in they have to enable both.

So when 
```
buildUponDefaultConfig = true
allRules = true
```
Are set, I expect to use default configuration for stable checks, together with experimental checks.

Let me know if I misunderstood, I just subscribed to this issue because of the title :)

**Comment by cortinico** [comment]
> `allRules` enables all rules on top of using the default config (which seems to be the code (`AllRulesConfig`) anyway.

IMHO that should be the preferred behaviour.
This is effectively a bug and should be fixed.

> allRules -> Enables the experimental, upcoming new rules.

@mustafaozhan There are no 'experimental' rules though. The `allRules` should just result in you enabling all the rules that are ever discovered by detekt. `buildUponDefaultConfig` will instead bring all the custom config keys that are in the default config.

Also, I believe that the semantic of:

```
buildUponDefaultConfig = true
allRules = true
```

Should result in having `allRules` being applied with a **higher** priority. In the sense that even if a rule is disabled in the default config, it should be enabled. We should probably fire a warning in this case maybe? 🤔

**Comment by TWiStErRob** [comment]
Agree with @mustafaozhan, he summed it up very nicely. No warnings necessary, because it's the wanted behaviour. I expected point 1 in OP too. `allRules` = "overlay a config where everything is `active: true`", and as you said this one should be applied last.

**Comment by mustafaozhan** [comment]
Yea I think I got confused from the Android Studio plugin, I remember in the past there was experimental term in the config.

Anyway, I think none of the flag really should affect the other one (if we can do...).

What I understand from your comment is 
`allRules > buildUponDefaultConfig` so If I set `allRules` then `buildUponDefaultConfig` will be already set, so no need to set buildUponDefaultConfig again, then it seems warning will be useful.

**Comment by TWiStErRob** [comment]
`allRules` could imply `buildUponDefaultConfig` *by default*, but I would imagine that
```
allRules = true
buildUponDefaultConfig = false
```
may be also valid setup, albeit a bit confusing since defaults are not used, but that's what the configurator asked for.

---

If any combination of the flags could be valid, and even though one implies the other (by default if not explicitly set (impl: convention property value)). Setting both to true would just make reading the Gradle configuration easier, if the author chooses to be explicit. Setting both doesn't change the behaviour and doesn't cause any problems, so there is no need for a warning.

**Comment by TWiStErRob** [comment]
What's the next step here?

Is this a candidate for https://github.com/detekt/detekt/milestone/42?
or Can we do it in 1.x?
or This is a won't fix, because ...?

**Comment by BraisGabin** [comment]
This should be doable on 1.x. To me it doesn't have a high priority but it's up to grab by anyone.
